File: frontend/src/users/Email.ts
export class Email {
    private static readonly EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

    private constructor(public readonly value: string) {}

    static isValid(email: string): boolean {
        return this.EMAIL_REGEX.test(email);
    }

    static of(email: string): Email {
        if (!this.isValid(email)) {
            throw new Error("Invalid email format");
        }
        return new Email(email);
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/types/Email.kt
package com.bellako.kiwi.types

@JvmInline
value class Email private constructor(val value: String) {
    companion object {
        private val EMAIL_REGEX = Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}$")

        fun isValid(email: String): Boolean {
            return EMAIL_REGEX.matches(email)
        }

        fun of(email: String): Result<Email> {
            return if (isValid(email)) {
                Result.success(Email(email))
            } else {
                Result.failure(IllegalArgumentException("Invalid email format"))
            }
        }
    }
}